File: exams/final/q4/solution.py
def partition(arr, l, r):
    x, i = arr[r], l
    for j in range(l, r):
        if arr[j] <= x:
            arr[i], arr[j] = arr[j], arr[i]
            i += 1
    arr[i], arr[r] = arr[r], arr[i]
    return i
 
def quickselect(arr, k, l, r):
    if k > 0 and k <= r - l + 1:
        index = partition(arr, l, r)
        if index-l == k-1:
            return arr[index]
        elif index - l > k - 1: # left subarray
            return quickselect(arr, k, l, index - 1)
        else: # right subarray
            return quickselect(arr, k-index+l-1, index+1, r)

def quicksort(arr, l, r):
    if l < r:
        p = partition(arr, l, r)
        quicksort(arr, l, p-1)
        quicksort(arr, p+1, r)

arr = [3,1,5,2,8,12,4]
k = 4
kthSmallest = quickselect(arr, k, 0, len(arr)-1)
print(kthSmallest)
quicksort(arr, 0, k)
print(arr)

File: exams/final/q4/Solution.java
import java.util.Arrays;

public class Solution {
    public static Integer partition(int[] arr, int l, int r) {
        int x = arr[r];
        int i = l;
        for (int j = l; j < r; j++) {
            if (arr[j] <= x) {
                int tmp = arr[i];
                arr[i] = arr[j];
                arr[j] = tmp;
                i += 1;
            }
        }
        int tmp = arr[i];
        arr[i] = arr[r];
        arr[r] = tmp;
        return i;
    }

    // return pivot = kthSmallest, move values less than pivot before pivot
    public static Integer quickselect(int[] arr, int k, int l, int r) {
        if (k > 0 && k <= r - l + 1) {
            int p = partition(arr, l, r);
            if (p - l == k - 1)
                return arr[p];
            else if (p - l > k - 1)
                return quickselect(arr, k, l, p - 1);
            else
                return quickselect(arr, k-p+l-1, p+1, r);
        }
        return -1;
    }

    // sort values between l and r
    public static void quicksort(int[] arr, int l, int r) {
        if (l < r) {
            int p = partition(arr, l, r);
            quicksort(arr, l, p - 1);
            quicksort(arr, p + 1, r);
        }
    }

    public static void main(String[] args) {
        int[] arr = {3,1,5,2,8,12,4};
        int k = 4;
        int kthSmallest = quickselect(arr, k, 0, arr.length - 1);
        quicksort(arr, 0, k);
        System.out.println(kthSmallest);
        System.out.println(Arrays.toString(arr));
    }
}
